x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/huggingface/response/HuggingFaceRerankResponseEntity.java

```diff
@@ -0,0 +1,130 @@
+
+import java.io.IOException;
+import java.util.Comparator;
+import java.util.List;
+
+import static org.elasticsearch.common.xcontent.XContentParserUtils.ensureExpectedToken;
+import static org.elasticsearch.common.xcontent.XContentParserUtils.parseList;
+import static org.elasticsearch.core.Strings.format;
+import static org.elasticsearch.xpack.inference.external.response.XContentUtils.moveToFirstToken;
+
+public class HuggingFaceRerankResponseEntity {
+
+    private static final String FAILED_TO_FIND_FIELD_TEMPLATE = "Failed to find required field [%s] in Hugging Face rerank response";
+    private static final String INVALID_ID_FIELD_FORMAT_TEMPLATE = "Expected numeric value for record ID field in Hugging Face rerank ";
+
+    /**
+     * Parses the Hugging Face rerank response.
+
+     * For a request like:
+     *
+     * <pre>
+     *     <code>
+     *         {
+     *              "input": ["luke", "like", "leia", "chewy","r2d2", "star", "wars"],
+     *              "query": "star wars main character",
+     *              "return_documents": true,
+     *              "top_n": 1
+     *          }
+     *     </code>
+     * </pre>
+
+     * The response would look like:
+
+     * <pre>
+     *     <code>
+     *         {
+     *              "rerank": [
+     *                  {
+     *                       "index": 5,
+     *                       "relevance_score": -0.06920313,
+     *                       "text": "star"
+     *                   }
+     *               ]
+     *          }
+     *     </code>
+     * </pre>
+     */
+
+    public static RankedDocsResults fromResponse(HuggingFaceRerankRequest request, HttpResult response) throws IOException {
+        var parserConfig = XContentParserConfiguration.EMPTY.withDeprecationHandler(LoggingDeprecationHandler.INSTANCE);
+
+        try (XContentParser jsonParser = XContentFactory.xContent(XContentType.JSON).createParser(parserConfig, response.body())) {
+            moveToFirstToken(jsonParser);
+
+            XContentParser.Token token = jsonParser.currentToken();
+            ensureExpectedToken(XContentParser.Token.START_ARRAY, token, jsonParser);
+
+            var rankedDocs = doParse(jsonParser);
+            var rankedDocsByRelevanceStream = rankedDocs.stream()
+                .sorted(Comparator.comparingDouble(RankedDocsResults.RankedDoc::relevanceScore).reversed());
+            var rankedDocStreamTopN = request.getTopN() == null
+                ? rankedDocsByRelevanceStream
+                : rankedDocsByRelevanceStream.limit(request.getTopN());
+            return new RankedDocsResults(rankedDocStreamTopN.toList());
+        }
+    }
+
+    private static List<RankedDocsResults.RankedDoc> doParse(XContentParser parser) throws IOException {
+        return parseList(parser, (listParser, index) -> {
+            var parsedRankedDoc = HuggingFaceRerankResponseEntity.RankedDocEntry.parse(parser);
+
+            if (parsedRankedDoc.id == null) {
```

Comment:
I believe `declare*()` requires that the result be non-null. So I think we can remove these if-blocks to check for non-null. Can we create a test to ensure that null is not valid?